### ISSUE TITLE
Difference between TFLite and USB Accelerator

### ISSUE THREAD

**Issue by denissuh** [google-coral/edgetpu#230]
Hi,

I'm using Resnet18 from torchvision and doing inference on EdgeTPU.
In order to do so, I'm converting the model as following: Pytorch -> Onnx - > Keras -> TFLite -> EdgeTPU.
I compared the output of each step:

**torch predictions vs onnxruntime predictions**
	**max**: 5.7220458984375e-06
	**std**: 1.8011219253821764e-06
	**min**: -7.331371307373047e-06

**torch predictions vs keras predictions**
	**max**: 1.4781951904296875e-05
	**std**: 4.7468447519349866e-06
	**min**: -1.5974044799804688e-05

**keras predictions vs tflite predictions**
	**max**: 0.8791558966040611
	**std**: 0.27655993514495225
	**min**: -0.8303738981485367

**tflite predictions vs edge predictions**
	**max**: 0.484074205160141
	**std**: 0.13129207757440303
	**min**: -0.36305565387010574

As we can see there is a small difference between torch, onnx and keras .
There is a more significant difference between keras and tflite, but this can be explained by post training quantization.
Also, there is significant difference between tflite and EdgeTPU, which I don't understand.
Why EdgeTPU compiler creates a difference between tflite and EdgeTPU ?

Thank you in advance,
Denis

**Code**:
```Python
from pathlib import Path
import subprocess

import numpy as np

import torch
from torchvision.models import resnet18
from torchvision.datasets import CIFAR10
from torchvision import transforms

import onnx
import onnxruntime
from onnx2keras import onnx_to_keras

import tensorflow as tf
import tflite_runtime.interpreter as tflite

########################################################################################################################
# load cifar dataset
########################################################################################################################
test_transform = transforms.Compose([transforms.Scale(256),
                                     transforms.CenterCrop(224),
                                     transforms.ToTensor(),
                                     transforms.Normalize(mean=[0.485, 0.456, 0.406],
                                                          std=[0.229, 0.224, 0.225])
                                     ])

test_set = CIFAR10('./data',
                   download=False,
                   train=False,
                   transform=test_transform)
test_loader = torch.utils.data.DataLoader(test_set,
                                          batch_size=1,
                                          shuffle=False,
                                          num_workers=0)

test_iterator = iter(test_loader)
test_img,  _ = test_iterator.next()

########################################################################################################################
# load resnet18 model
########################################################################################################################
model_type = "resnet18"

# load resnet 18 model
model = resnet18(pretrained=True)
model.eval()
with torch.no_grad():
    predictions = model(test_img)

########################################################################################################################
# convert torch to onnx and compare predictions to torch
########################################################################################################################
torch.onnx.export(model, test_img, f"{model_type}.onnx", input_names=['input'])

onnxruntime_session = onnxruntime.InferenceSession(f"{model_type}.onnx")
onnx_inputs = {onnxruntime_session.get_inputs()[0].name: test_img.numpy()}
onnx_predictions = onnxruntime_session.run(None, onnx_inputs)

print("torch predictions vs onnxruntime predictions")
print(f"max: {np.max(predictions.numpy() - onnx_predictions[0])}")
print(f"std: {np.std(predictions.numpy() - onnx_predictions[0])}")
print(f"min: {np.min(predictions.numpy() - onnx_predictions[0])}")

########################################################################################################################
# convert onnx to keras and compare predictions to torch
########################################################################################################################
onnx_model = onnx.load(f"{model_type}.onnx")
keras_model = onnx_to_keras(onnx_model=onnx_model,
                            input_names=['input'],
                            change_ordering=True,
                            verbose=False)
keras_model.save(f"{model_type}.h5")

keras_model = tf.keras.models.load_model(f"{model_type}.h5")
keras_predictions = keras_model.predict(test_img.numpy().transpose([0, 2, 3, 1]))

print("torch predictions vs keras predictions")
print(f"max: {np.max(predictions.numpy() - keras_predictions)}")
print(f"std: {np.std(predictions.numpy() - keras_predictions)}")
print(f"min: {np.min(predictions.numpy() - keras_predictions)}")

########################################################################################################################
# convert keras to tflite and compare predictions to torch
########################################################################################################################
converter_keras = tf.lite.TFLiteConverter.from_keras_model_file(f"{model_type}.h5")
converter_keras.optimizations = [tf.lite.Optimize.DEFAULT]
converter_keras.target_spec.supported_ops = [tf.lite.OpsSet.TFLITE_BUILTINS_INT8]
converter_keras.inference_input_type = tf.uint8
converter_keras.inference_output_type = tf.uint8

def representative_sspd_data_gen():
    for _ in range(100):
        img, label = test_iterator.next()
        yield [img.numpy().transpose([0, 2, 3, 1])]

converter_keras.representative_dataset = representative_sspd_data_gen
tflite_qunatized_model = converter_keras.convert()

with tf.io.gfile.GFile(str(f"{model_type}.tflite"), 'wb') as f:
    f.write(tflite_qunatized_model)

tflite_path = f"{model_type}.tflite"
tflite_quantized_interpreter = tflite.Interpreter(model_path=tflite_path)
tflite_quantized_interpreter.allocate_tensors()

input_details_tflite = tflite_quantized_interpreter.get_input_details()[0]
input_scale_tflite, input_zero_point_tflite = input_details_tflite["quantization"]
test_image_tflite = test_img.numpy().transpose([0, 2, 3, 1])
test_image_tflite = test_image_tflite / input_scale_tflite + input_zero_point_tflite
test_image_tflite = test_image_tflite.astype(input_details_tflite["dtype"])
tflite_quantized_interpreter.set_tensor(input_details_tflite['index'], test_image_tflite)

tflite_quantized_interpreter.invoke()

output_details_tflite = tflite_quantized_interpreter.get_output_details()[0]
output_tflite = tflite_quantized_interpreter.get_tensor(output_details_tflite['index'])
scale_tflite, zero_point_tflite = output_details_tflite['quantization']
output_tflite = float(scale_tflite) * (output_tflite.astype(np.float) - float(zero_point_tflite))

print("keras predictions vs tflite predictions")
print(f"max: {np.max(keras_predictions - output_tflite)}")
print(f"std: {np.std(keras_predictions - output_tflite)}")
print(f"min: {np.min(keras_predictions - output_tflite)}")

########################################################################################################################
# convert tflite to edge tpu and compare predictions to tflite
########################################################################################################################

print("compile to edge TPU")
outfile_wsl_path = "/mnt/" + Path(Path.cwd(), f"{model_type}.tflite").as_posix().lower().replace(":", "")

subprocess.call(["ubuntu1804.exe",
                 "run",
                 "edgetpu_compiler",
                 outfile_wsl_path,
                 "--show_operations"])

edge_tpu_path = f"{model_type}_edgetpu.tflite"
tflite_quantized_interpreter_edge = tflite.Interpreter(model_path=edge_tpu_path,
                                                       experimental_delegates=[tflite.load_delegate('edgetpu.dll')])
tflite_quantized_interpreter_edge.allocate_tensors()

input_details_edge = tflite_quantized_interpreter_edge.get_input_details()[0]
input_scale_edge, input_zero_point_edge = input_details_edge["quantization"]
test_image_edge = test_img.numpy().transpose([0, 2, 3, 1])
test_image_edge = test_image_edge / input_scale_edge + input_zero_point_edge
test_image_edge = test_image_edge.astype(input_details_edge["dtype"])
tflite_quantized_interpreter_edge.set_tensor(input_details_edge['index'], test_image_edge)

tflite_quantized_interpreter_edge.invoke()

output_details_edge = tflite_quantized_interpreter_edge.get_output_details()[0]
output_edge = tflite_quantized_interpreter_edge.get_tensor(output_details_edge['index'])
scale_edge, zero_point_edge = output_details_edge['quantization']
output_edge = float(scale_edge) * (output_edge.astype(np.float) - float(zero_point_edge))

print("tflite predictions vs edge predictions")
print(f"max: {np.max(output_tflite - output_edge)}")
print(f"std: {np.std(output_tflite - output_edge)}")
print(f"min: {np.min(output_tflite - output_edge)}")

```

**Comment by Namburger** [comment]
Interesting, I'm not sure if we  ever tested this pipeline, there are a couple things I can think of here:

1) Model created from pytorch and then converts to keras -> tflite, I guess technically it's possible, but I imagine the implementation difference can play a role here. 

2) This trick:
```
subprocess.call(["ubuntu1804.exe",
                 "run",
                 "edgetpu_compiler",
                 outfile_wsl_path,
                 "--show_operations"])
```
Again, I guess it's also technically do able, but I would definitely investigate this first. The executable was written and compiled specifically for linux, so I wouldn't trust having another variable in this big pipeline. If you don't mind paste the quantized tflite model here, I can try compiling it first on my side on an updated compiler before you evaluate it.

**Comment by denissuh** [comment]
1. Could you please explain why implementation difference can play a role here ? I understand that there is a difference between PyTorch and TensorFlow, but once I moved to the TensorFlow EcoSystem, why the EdgeTPU compiler creates a difference between *.tflite and *_edge.tflite models ?

2. Though I attached a windows code, the difference between the models exists in Linux too.

Quantized tflite model attached [resnet18.zip](https://github.com/google-coral/edgetpu/files/5365757/resnet18.zip), it would be great if you could check it on your side, thank you !

**Comment by Namburger** [comment]
@denissuh 
To be clear, 1 is just my speculation, since the edgetpu architecture is fundamentally different than a normal CPU, the compiler should be aware of a normal tflite model, however your pipeline is much different than what we expected. 

For 2, I compiled your model on my linux with a more updated compiler version, could you test this?
[resnet18.tar.gz](https://github.com/google-coral/edgetpu/files/5365880/resnet18.tar.gz)

Would you consider using our [resnet50](https://www.tensorflow.org/api_docs/python/tf/keras/applications/ResNet50) implementation?

**Comment by denissuh** [comment]
Hi, 

- My project has a more complicated pipeline, thus unfortunately I cannot use tf.keras.applications.ResNet50 at the moment. 

- Unfortunately, the difference between tflite and EdgeTPU still exists with the model that you've compiled.
I wondering if it can be a hardware issue. Is there a chance you could check the models with your USB Accelerator ?

- Another related issue, I found that the layer that affects the most the difference between tflite and edge models is batch normalization. Specifically, when if I remove the batch normalization layer the std between tflite and edge outputs is 0.0023. 
As far as I understand TFLite folds batch normalization into the conv layer. Are there any known constraints when using batch normalization ? 
 
- In general, are there any constraints (beside that the model should use only the operations that are supported by the Edge TPU) when building TensorFlow model for EdgeTPU ?
For example, is there some constraint for the numbers of filters should be power of 2 ? Is there some constraint for maximum image size ? 

Thank you in advance,
Denis

**Comment by denissuh** [comment]
Hi, 

- My project has a more complicated pipeline, thus unfortunately I cannot use tf.keras.applications.ResNet50 at the moment. 

- Unfortunately, the difference between tflite and EdgeTPU still exists with the model that you've compiled.
I wondering if it can be a hardware issue. Is there a chance you could check the models with your USB Accelerator ?

- Another related issue, I found that the layer that affects the most the difference between tflite and edge models is batch normalization. Specifically, if I remove the batch normalization layer the std between tflite and edge outputs is 0.0023. 
As far as I understand TFLite folds batch normalization into the conv layer. Are there any known constraints when using batch normalization ? 
 
- In general, are there any constraints (besides that the model should use only the operations that are supported by the Edge TPU) when building TensorFlow model for EdgeTPU ?
For example, is there some constraint for the numbers of filters should be power of 2 ? Is there some constraint for maximum image size ? 

Thank you in advance,
Denis

**Comment by Naveen-Dodda** [comment]
Hello Denis, 

There are no such constraints while building a model for edgetpu . I am attaching a new compiled model, could you check if predictions are still different with this model. Its interesting to know that Can you share us both .tflite models with and without batch norm ? 

[resnet18_edgetpu.tar.gz](https://github.com/google-coral/edgetpu/files/5448162/resnet18_edgetpu.tar.gz)

Thanks
